rfcs/proposed/host_backend_histogram/README.md

```diff
@@ -0,0 +1,73 @@
+
+### Code Reuse
+Our goal here is to make something maintainable and to reuse as much as we can which already exists and has been reviewed within oneDPL. With everything else, this must be balanced with performance considerations.
+
+### unseq Backend
+As mentioned above, histogram looks to be a memory bandwidth-dependent algorithm. This may limit the benefit achievable from vector instructions as they provide assistance mostly in speeding up computation. Vector operations in this case also compound our issue of race conditions, multiplying the number of concurrent lines of execution by the vector length. The advantage we get from vectorization of the increment operation or the lookup into the output histogram may not provide much benefit, especially when we account for the extra memory footprint required or synchronization required to overcome the race conditions which we add from the additional concurrent streams of execution. It may make sense to decline to add vectorized operations within histogram depending on the implementation used, and based on performance results.
+
+## Existing Patterns
+
+### count_if
+`histogram` is similar to `count_if` in that it conditionally increments a number of counters based upon the data in a sequence. `count_if` returns a scalar-typed value and doesn't provide any function to modify the variable being incremented. Using `count_if` without significant modification would require us to loop through the entire sequence for each output bin in the histogram. From a memory bandwidth perspective, this is untenable. Similarly, using a `histogram` pattern to implement `count_if` is unlikely to provide a well-performing result in the end, as contention should be far higher, and `reduce` is a very well-matched pattern performance-wise.
+
+### parallel_for
+`parallel_for` is an interesting pattern in that it is very generic and embarrassingly parallel. This is close to what we need for `histogram`. However, we cannot simply use it without any added infrastructure. If we were to just use `parallel_for` alone, there would be a race condition between threads when incrementing the values in the output histogram. We should be able to use `parallel_for` as a building block for our implementation, but it requires some way to synchronize and accumulate between threads.
+
+## Proposal
+I believe there are two competing options for `histogram`, which may both have utility in the final implementation depending on the use case.
+
+### Implementation One (Embarrassingly Parallel)
+This method uses temporary storage and a pair of embarrassingly parallel `parallel_for` loops to accomplish the `histogram`.
+1) Determine the number of threads that we will use, perhaps adding some method to do this generically based on the backend.
+2) Create temporary data for the number of threads minus one copy of the histogram output sequence. Thread zero can use the user-provided output data.
```

Comment:
Determining/assuming a certain number of threads is not necessary, and should be avoided for the sake of parallel composability. We need a sort of thread-local storage, maybe the one that is accessible by other threads for data aggregation. It certainly can be done without knowing the number of threads.

Reply:
There is some precedent for utilizing the maximum number of threads for accumulation in the OpenMP backend with `transform_reduce` [(link)](https://github.com/oneapi-src/oneDPL/blob/bdc037b9335cc18e29c935cef75204272e5d6341/include/oneapi/dpl/pstl/omp/parallel_transform_reduce.h#L42).  This is of course not to say it is correct or best.

Reply:
On the OpenMP side, perhaps we shouldn't be using `parallel_for` but rather extending some of its building blocks for use by `histogram` and `parallel_for`, and with more control over `__chunk_partitioner` and the parallel tasks / loop is what we need, and where we can understand our number of local temporary copies / threads which will use them.

For TBB, I'm more in the dark since that is more recursively decomposing into chunks , but looking further...

Reply:
For TBB, does it make sense to do something similar to [this code](https://github.com/oneapi-src/oneDPL/blob/bdc037b9335cc18e29c935cef75204272e5d6341/include/oneapi/dpl/pstl/parallel_backend_tbb.h#L359-L366)?

The plan would be to Isolate then get the max concurrency of the task arena and allocating temporaries based on that. Then using `current_thread_index()` within a function brick which is decomposed directly or via a `parallel_for` pattern followed by a launch of tasks to accumulate those into the output.

It seems like this is not thinking very "tbb-like", because we would be interacting directly with thread indices and number of threads, which is discouraged.  However, I this case I think its important to control the allocation while not sacrificing the ability to use a small grain size for work stealing and load balance.  

Is there some more tbb-way of doing this which would have a similar control on footprint?

Reply:
Aha, this is what I think I was missing... 
https://github.com/uxlfoundation/oneAPI-spec/blob/main/source/elements/oneTBB/source/thread_local_storage/enumerable_thread_specific_cls.rst

It may make it difficult to understand where the appropriate heuristic for choosing between algorithms, but I think this is the "tbb way" of doing this which I was missing.

Reply:
Yes, in TBB there is `enumerable_thread_specific`. But even without TBB, something like this could be done with e.g. a hash map protected by a lock. A thread ID/native handle would be the key to the map; and if it is guaranteed that insertions do not invalidate iterators/references to the map nodes, the lock would only need to be taken for getting such a reference, which could then be used to update the content without data races.

I am not sure if it's better to have a generic implementation like I outlined above, or if a special "interface" with backend-specific implementations is necessary for performance reasons.